### ISSUE TITLE
Merge global forecasting capabilities from `prob_model_global`

### ISSUE THREAD

**Issue by joeloskarsson** [mllam/neural-lam#63]
Global forecasting can be viewed as a special case of LAM, with no boundary conditions to consider. Therefore it is not a hard task to allow for global forecasting within the Neural-LAM framework. This has the benefit of being able to use any models in here also for global forecasting.

Global forecasting, as done in [Oskarsson et al. (2024)](https://arxiv.org/abs/2406.04759), is currently implemented on the branch `prob_model_global`. This issue is to keep track of merging of the `prob_model_global` branch into main.

The main changes in this branch include:

- [ ] Add script for creating of global forcing https://github.com/mllam/neural-lam/blob/f68a5132a1d6aafcd5695705762d08ce29cbb777/create_global_forcing.py#L26 This will likely be taken care of through #34, rather than merging `create_global_forcing.py` directly.
- [ ] Add script for creating global static grid features https://github.com/mllam/neural-lam/blob/f68a5132a1d6aafcd5695705762d08ce29cbb777/create_global_grid_features.py#L23 Also likely better handled in #34.
- [ ] Add script for creating global mesh graphs (multiscale and hierarchical) https://github.com/mllam/neural-lam/blob/f68a5132a1d6aafcd5695705762d08ce29cbb777/create_global_mesh.py#L128 Will see if this should sit in this repo or in [weather-model-graphs](https://github.com/mllam/weather-model-graphs), and to what extent we keep the graphcast dependency for graph generation.
- [ ] Allow for 3d graph plotting of global graphs https://github.com/mllam/neural-lam/blob/f68a5132a1d6aafcd5695705762d08ce29cbb777/plot_global_graph.py#L99 It is perhaps better to first move the 3d plotting functionality to [weather-model-graphs](https://github.com/mllam/weather-model-graphs) and do this there. See point above.
- [ ] Enable generating variable weights based on pressure levels https://github.com/mllam/neural-lam/blob/f68a5132a1d6aafcd5695705762d08ce29cbb777/create_parameter_weights.py#L53 This is better handled by defining more general variable weighting setup where one can either set these weights manually per-variable in config files or use some pre-defined strategies.
- [ ] Add ERA5 dataset class https://github.com/mllam/neural-lam/blob/f68a5132a1d6aafcd5695705762d08ce29cbb777/neural_lam/era5_dataset.py#L13 This should not be added as a separate new class, as all functionality in there should be possible through the generic dataset created in #54. But it needs to be checked that the result of #54 will work with global ERA5 data.
- [ ] Add functionality to save forecasts to zarr+xarray for later plotting or evaluation https://github.com/mllam/neural-lam/blob/f68a5132a1d6aafcd5695705762d08ce29cbb777/neural_lam/forecast_to_xarr.py#L199 
- [ ] Add grid weighting (spatial points weighted differently) to loss and metrics https://github.com/mllam/neural-lam/blob/f68a5132a1d6aafcd5695705762d08ce29cbb777/neural_lam/metrics.py#L41 
- [ ] Make boundary forcing optional when rolling out forecasts https://github.com/mllam/neural-lam/blob/f68a5132a1d6aafcd5695705762d08ce29cbb777/neural_lam/models/ar_model.py#L150
- [ ] Plot example predictions for specific variables and lead times also during validation steps https://github.com/mllam/neural-lam/blob/f68a5132a1d6aafcd5695705762d08ce29cbb777/neural_lam/models/base_graph_model.py#L183
- [ ] Allow for plotting of global data https://github.com/mllam/neural-lam/blob/f68a5132a1d6aafcd5695705762d08ce29cbb777/neural_lam/vis.py#L79 Plotting functionality should be based on arbitrary projections rather than hard-coded to MEPS case. This should be done already by #54, but it needs to be checked that this works also for global forecasts.

**Comment by aadilrizwan** [comment]
Hi @joeloskarsson ,I noticed that ML-LAM is listed among the GSoC 2026 organizations, and I am keen to contribute. I would like to confirm whether this issue is open for contributors and if I may begin working on it as a potential GSoC proposal.

**Comment by joeloskarsson** [comment]
@InFiNiTy0639 yes, this issue is highly related to project idea number 4. Please see https://github.com/mllam/neural-lam/wiki/GSoC-ideas#4-global-weather-forecasting for some more details.

**Comment by yuvraajnarula** [comment]
Hi @joeloskarsson,

I’ve been reviewing the code related to this issue, particularly `ARModel` and `metrics.py` , to better understand how global forecasting is expected to integrate with the current LAM-specific rollout logic.

From what I can see:
- Boundary forcing is tightly embedded in `unroll_prediction()`, which may require some decoupling for a clean global implementation.
- The metrics layer already appears fairly general and likely needs minimal changes.

Before diving deeper, I wanted to clarify the intended scope and timing of this issue:

- Is #63 currently open for incremental contributions?
- Or is it expected to be addressed mainly during the GSoC period?
- Are there specific sub-parts (e.g., boundary abstraction, mesh handling, dataset integration) that you would consider safe and useful to work on now?

I’d appreciate any guidance on where contributions would be most helpful at this stage.

**Comment by nehamanoj1105** [comment]
Hi @joeloskarsson ,

I’ve opened a small PR (#211) that makes boundary forcing optional in ARModel, as a first step toward clean global forecasting support.

This focuses only on decoupling boundary overwrite logic during rollout.

Happy to refine further if needed.

**Comment by selcuksntrk** [comment]
Hi @joeloskarsson @observingClouds @leifdenby,

I'm Selcuk Senturk, MSc Computer Science student and an AI/ML Engineer in IBM. I'm interested in contributing to the global weather forecasting extension as part of GSoC 2026 (175 hours).

My background connects to this project on multiple levels. My BSc in Physics means graph construction on spherical domains, coordinate systems, and the mathematics underlying GNN message passing are familiar ground, not just as ML tools but as physical representations. I work daily with PyTorch in production systems, and I've worked with graph-based deep learning architectures in my research and professional work. I'm also comfortable with data pipelines from working with structured scientific datasets.

The core challenge here, constructing a well-behaved global graph that handles the periodicity of longitude and the polar singularities is the kind of problem I find technically interesting.

I would like to discuss further if I'm a good candidate for this project under GSoC 2026 and how I plan to contribute.

Resume: [Drive](https://drive.google.com/file/d/12NCptLXY1FZIreGX6pGKuQFi17XRQjTQ/view?usp=sharing)
LinkedIn: [selcuksntrk](https://www.linkedin.com/in/selcuksntrk/)

**Comment by joeloskarsson** [comment]
@yuvraajnarula we mainly view this as something that could be worked on throughout gsoc rather than now. But I will not stop any contributions already now.

@selcuksntrk Hi! Feel free to also join the slack and discuss further in the project channel there.

**Comment by Mani212005** [comment]
Hi @joeloskarsson,

I'm interested in contributing to this issue as part of GSoC 2026. I've reviewed the branch and the related issues, and I saw that the forecast_to_xarr.py functionality hasn't been addressed yet in any open PRs.

Would it be useful for me to work on porting this functionality to main, ensuring it integrates cleanly with the generic dataset work from #54? I'd open a draft PR early to get feedback on the direction before going too far.
Thank you!

**Comment by AR10129** [comment]
@joeloskarsson 
I'd like to work on the **grid area weighting sub-task** (`metrics.py:41` in `prob_model_global`) as part of GSoC 2026. Since `BaseDatastore.get_xy()` returns projected coordinates, lat/lon can be recovered via the existing `coords_projection()` without touching either backend. 

Planned changes:
1. Add `get_latitudes()` to `BaseDatastore` using `get_xy()` + `coords_projection()`
2. Add `area_weighting: bool = False` to `TrainingConfig`
3. Compute normalized `cos(lat)` weights in `ARModel.__init__`, register as buffer
4. Thread `grid_weights` through training/validation/test loss and metric calls

Would you prefer the cartopy transform abstracted behind `get_latitudes()` on `BaseDatastore`, or done directly inside `ARModel`?

Opened a draft PR here: #368

**Comment by Palakgokhru** [comment]
Hi! I’m interested in working on integrating global forecasting
capabilities into Neural-LAM for GSoC 2026.

I’ve been exploring the repository and understanding how the
current LAM pipeline works (datasets, graph generation and
training modules).

From reading the issue, my understanding is that the goal is to
merge functionality from the `prob_model_global` branch including:

• global forcing generation
• global grid feature generation
• mesh creation for global forecasting
• dataset integration (ERA5)
• global visualization and evaluation

As a first step, I plan to:

1. Study the scripts mentioned in the issue.
2. Refactor them into reusable modules.
3. Integrate global mesh generation into the current pipeline.

Does this approach sound reasonable?

**Comment by opeblow** [comment]
Hi @joeloskarsson, I'm interested in working on global forecasting support for GSoC 2026. I've studied the current create_graph.py implementation in detail and identified the three core areas that need to change for spherical support , the coordinate system assumption, flat surface dot placement using meshgrid, and Euclidean distance calculation that breaks on a sphere. I have a background in ML systems engineering including graph-based environments and I'm comfortable with PyTorch and PyTorch Geometric. Would you be open to discussing the scope of what a GSoC contributor could realistically deliver in 175 hours? I'm drafting a proposal now and would value your input before March 16.

**Comment by Franklingarci** [comment]
Hi @joeloskarsson, I’m interested in contributing to this project 4. Since it’s meant for GSoC, I was wondering if there are similar issues that I could work on ?

**Comment by joeloskarsson** [comment]
@Franklingarci @opeblow @Palakgokhru For questions on writing a proposal around this for gsoc please ask in the corresponding slack-channel, to avoid too broad discussions in this thread.

**Comment by ayushhbasu** [comment]
Hi @joeloskarsson,

While exploring the requirements for global graph construction, I noticed that BaseDatastore already exposes geographic coordinates via get_lat_lon(), which seems like a natural foundation for introducing global-aware geometry utilities.

I’m considering a small, incremental PR that would add:

- Latitude-based area weights (cos(lat)-based normalization) for physically consistent area-weighted metrics
- Spherical coordinate conversion from (lon, lat) to (x, y, z) for geometry-aware distance computations
- A minimal, optional graph builder interface (e.g., KNN over spherical coordinates), designed to remain backend-agnostic

The intention is not to introduce a full global graph pipeline at this stage, but rather to establish a clean, reusable geometry layer that can support future work—either native implementations or integration with external libraries such as weather-model-graphs.

Would this be a reasonable first step toward enabling global support in a modular way?

Happy to open a draft PR if this aligns.

**Comment by ayushhbasu** [comment]
Thanks @RajdeepKushwaha5 . I've opened a draft PR (#473) with the geometry utilities:

- get_area_weights() and get_cartesian_coords() added to BaseDatastore in neural_lam/datastores/base.py
- neural_lam/geometry.py with lat_lon_to_cartesian() and get_area_weights()
- neural_lam/graph_builder.py with build_spherical_knn_graph() (scipy backend)
- Tests will be added after #444 merges (using GlobalDummyDatastore)

The PR is marked as draft and blocked on #444 for full test coverage. Happy to adjust the location of these utilities if anyone has strong preferences — I kept the graph builder separate from BaseDatastore to avoid bloat.

**Comment by Yusufiabichi** [comment]
Hi @joeloskarsson, I've been reading through `ar_model.py` to understand how global forecasting would fit in.

From what I see, `optional_boundary_forcing()` already makes the rollout logic fairly clean, and if `border_mask=None` the model seems to naturally fall back to fully autoregressive prediction. So it looks like ARModel itself might already be close to global-ready.

I'm wondering if the bigger challenges are expected more on the data/graph generation side rather than the rollout logic itself?

Still getting familiar with the codebase, so any correction is welcome.

**Comment by arzoo0511** [comment]
Hi @joeloskarsson ,

I’ve been analyzing the global forecasting requirements, and it seems the main bottleneck is not ARModel (which already supports boundary-free rollout), but rather the lack of a clean spherical geometry + graph construction layer.
I’m currently working on a modular approach that separates spherical coordinate utilities and the backend-agnostic global graph builder .my  goal is to make global support plug into the existing pipeline without introducing special-case logic.
I’d like to open a draft PR focused only on the geometry and minimal graph builder layer as a first step. Would this align with the intended direction?
